src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,279 @@
+    if (r == -ENOENT) {
+      break; // bucket is deleted, we should delete the v1 notification
+    }
+    if (r < 0) {
+      std::string s = "failed to load the bucket from " + obj.to_str()
+                    + " with: " + cpp_strerror(r);
+      rgw_clog_warn(rados, s);
+      return r;
+    }
+
+    if (bucket->get_marker() != marker) {
+      break;
+    }
+
+    bufferlist bl;
+    bucket_topics.encode(bl);
+    // same attrs ready to merge
+    rgw::sal::Attrs attrs = bucket->get_attrs();
+
+    if(bucket->get_attrs().contains(RGW_ATTR_BUCKET_NOTIFICATION)) {
+      // change only bucket notification attr
+      attrs[RGW_ATTR_BUCKET_NOTIFICATION] = std::move(bl);
+      r = bucket->merge_and_store_attrs(dpp, attrs, y);
+    } else {
+      bucket->get_attrs()[RGW_ATTR_BUCKET_NOTIFICATION] = std::move(bl);
+      r = bucket->put_info(dpp, false, real_time(), y);
+    }
+
+    if (r != -ECANCELED && r < 0) {
+      ldpp_dout(dpp, 1) << "ERROR: failed writing bucket instance info: " << cpp_strerror(-r) << dendl;
+      std::string s = "ERROR: failed writing bucket instance info: " + cpp_strerror(-r);
+      rgw_clog_warn(rados, s);
+      return r;
+    }
+  }
+
+  // delete v1 notification obj with Bucket::remove_topics()
+  r = rados_bucket.remove_topics(&bucket_topics_objv, y, dpp);
+  if (r == -ECANCELED) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    std::string s = "failed to remove migrated v1 bucket topics obj" + obj.to_str() + " with: " + cpp_strerror(-r);
+    rgw_clog_warn(rados, s);
+    return r;
+  }
+
+  return r;
+}
+
+// migrate topics for a given tenant
+int migrate_topics(const DoutPrefixProvider* dpp, optional_yield y,
+                   rgw::sal::RadosStore* driver,
+                   const rgw_raw_obj& topics_obj)
+{
+  // parse tenant name out of topics_obj "pubsub.{tenant}"
+  auto* rados = driver->getRados()->get_rados_handle();
+  std::string tenant;
+  const auto& topics_obj_oid = topics_obj.oid;
+  if (auto pos = topics_obj_oid.find(rgw::sal::pubsub_oid_prefix); pos != std::string::npos) {
+    tenant = topics_obj_oid.substr(std::string(rgw::sal::pubsub_oid_prefix).size());
+  } else {
+    std::string s = "failed to read tenant from name from oid: " + topics_obj_oid + " with: " + cpp_strerror(-EINVAL);
+    rgw_clog_warn(rados, s);
+    return -EINVAL;
+  }
+
+  // migrate the topics
+  rgw_pubsub_topics topics;
+  RGWObjVersionTracker topics_objv;
+  int r = driver->read_topics(tenant, topics, &topics_objv, y, dpp);
+  if (r == -ENOENT) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    std::string s = "failed to read v1 topics from " + topics_obj.to_str() + " error: " + cpp_strerror(-r);
+    rgw_clog_warn(rados, s);
+    return r;
+  }
+
+  constexpr bool exclusive = true; // don't overwrite any existing v2 metadata
+  for (const auto& [name, topic] : topics.topics) {
+    if (topic.name != topic.dest.arn_topic) {
```

Comment:
what does this mean? why can't we transition topics like this?

Reply:
These are auto generated topics and they are hidden from the user,
we don't use them anymore in v2

Reply:
what auto-generates such topics?

Reply:
In V1, whenever we create a bucket notification it generates these topics, correct @yuvalif?

Reply:
There is something that confuses me, how the auto generated topics have a linked bucket after migration, while the original doesn't.
original:
```
       {
            "user": "zone.user",
            "name": "fishtopic2",
            "dest": {
                "push_endpoint": "kafka://localhost",
                "push_endpoint_args": "Version=2010-03-31",
                "push_endpoint_topic": "fishtopic2",
                "stored_secret": false,
                "persistent": false,
                "time_to_live": "None",
                "max_retries": "None",
                "retry_sleep_duration": "None"
            },
            "arn": "arn:aws:sns:zg1::fishtopic2",
            "opaqueData": "",
            "policy": "",
            "subscribed_buckets": []
        },
```

auto generated:
```
        {
            "user": "zone.user",
            "name": "notif1_fishtopic2",
            "dest": {
                "push_endpoint": "kafka://localhost",
                "push_endpoint_args": "Version=2010-03-31",
                "push_endpoint_topic": "fishtopic2",
                "stored_secret": false,
                "persistent": false,
                "time_to_live": "None",
                "max_retries": "None",
                "retry_sleep_duration": "None"
            },
            "arn": "arn:aws:sns:zg1::fishtopic2",
            "opaqueData": "",
            "policy": "",
            "subscribed_buckets": [
                "fish"
            ]
        }
```

Reply:
> In V1, whenever we create a bucket notification it generates these topics, correct @yuvalif?

yes, in V1 we created this "internal" topic for each notification. this is because we had only a "topic" object that included things that were specific to notifications (e.g. filters).
anyway, this is something I'm really glad we got rid of :-)

Reply:
this 'internal topic' was separate from what we're calling the 'v1 notification' object which is per-bucket?

what value do those topics use for `arn_topic`?

can we guarantee that none of those internal topics are still in use?

Reply:
I presume that we don't need them anymore in v2, but my tests are failing because I didn't migrate them